aspects/aspects_test.go

```diff
@@ -865,27 +841,235 @@ func (s *aspectSuite) TestGetRulesAreSortedByParentage(c *C) {
 
-	value = nil
-	err = aspect.Get(databag, "foo", &value)
+	value, err = aspect.Get(databag, "foo")
 	c.Assert(err, IsNil)
 	// the leaf is replaced by a value read from a rule that is nested
 	c.Assert(value, DeepEquals, map[string]interface{}{"foo": map[string]interface{}{"bar": map[string]interface{}{"baz": "second"}}})
 
 	err = databag.Set("third", "third")
 	c.Assert(err, IsNil)
 
-	value = nil
-	err = aspect.Get(databag, "foo", &value)
+	value, err = aspect.Get(databag, "foo")
 	c.Assert(err, IsNil)
 	// lastly, it reads the value from "foo.bar.baz" the most nested entry
 	c.Assert(value, DeepEquals, map[string]interface{}{"foo": map[string]interface{}{"bar": map[string]interface{}{"baz": "third"}}})
 }
+
+func (s *aspectSuite) TestGetUnmatchedPlaceholderReturnsAll(c *C) {
+	databag := aspects.NewJSONDataBag()
+	aspectBundle, err := aspects.NewAspectBundle("acc", "bundle", map[string]interface{}{
+		"snaps": []map[string]string{
+			{"request": "snaps.{snap}", "storage": "snaps.{snap}"},
+		},
+	}, aspects.NewJSONSchema())
+	c.Assert(err, IsNil)
+	aspect := aspectBundle.Aspect("snaps")
+	c.Assert(aspect, NotNil)
+
+	err = databag.Set("snaps", map[string]interface{}{
+		"snapd": 1,
+		"foo": map[string]interface{}{
+			"bar": 2,
+		},
+	})
+	c.Assert(err, IsNil)
+
+	value, err := aspect.Get(databag, "snaps")
+	c.Assert(err, IsNil)
+	c.Assert(value, DeepEquals, map[string]interface{}{"snaps": map[string]interface{}{"snapd": float64(1), "foo": map[string]interface{}{"bar": float64(2)}}})
+}
+
+func (s *aspectSuite) TestGetUnmatchedPlaceholdersWithNestedValues(c *C) {
+	databag := aspects.NewJSONDataBag()
+	aspectBundle, err := aspects.NewAspectBundle("acc", "bundle", map[string]interface{}{
+		"statuses": []map[string]string{
+			{"request": "snaps.{snap}.status", "storage": "snaps.{snap}.status"},
+		},
+	}, aspects.NewJSONSchema())
+	c.Assert(err, IsNil)
+	asp := aspectBundle.Aspect("statuses")
+	c.Assert(asp, NotNil)
+
+	err = databag.Set("snaps", map[string]interface{}{
+		"snapd": map[string]interface{}{
+			"status": "active",
+		},
+		"foo": map[string]interface{}{
+			"version": 2,
+		},
+	})
+	c.Assert(err, IsNil)
+
+	value, err := asp.Get(databag, "snaps")
+	c.Assert(err, IsNil)
+	c.Assert(value, DeepEquals, map[string]interface{}{"snaps": map[string]interface{}{"snapd": map[string]interface{}{"status": "active"}}})
+}
+
+func (s *aspectSuite) TestGetSeveralUnmatchedPlaceholders(c *C) {
+	databag := aspects.NewJSONDataBag()
+	aspectBundle, err := aspects.NewAspectBundle("acc", "bundle", map[string]interface{}{
+		"foo": []map[string]string{
+			{"request": "a.{b}.c.{d}.e", "storage": "a.{b}.c.{d}.e"},
+		},
+	}, aspects.NewJSONSchema())
+	c.Assert(err, IsNil)
+	asp := aspectBundle.Aspect("foo")
+	c.Assert(asp, NotNil)
+
+	err = databag.Set("a", map[string]interface{}{
+		"b1": map[string]interface{}{
+			"c": map[string]interface{}{
+				// the request can be fulfilled here
+				"d1": map[string]interface{}{
+					"e": "end",
```

Comment:
do we need tests where there are other values here parallel to "e"? that would be ignored?

Reply:
It might be covered elsewhere but it's an easy win here so I added a value at e's level. Thank you